### ISSUE TITLE
[sharding] Use gRPC to access data on a remote shard

### ISSUE THREAD

**Issue by e-ivkov** [qdrant/qdrant#377]
Depends on: #373 

Alternative: use libp2p, seems too complex for our needs

**Comment by generall** [comment]
Based on discussion with @e-ivkov, I would like to add some details to the task and maybe split it into several more.
There are several challenge to solve and decisions to make. 

First of all, `RemoteShard` should be able to make requests to the remote services. There are two alternatives:

* it should make requestst itself
  * in this case - It should contain enough information to make a request - i.e. host, port, target shard id
  * in case of partitions moving - the will be a meta collection update operation which will re-create remote shards with updated information
* It uses some other abstraction to make requests
  * in this case - `RemoteShard` should have an access to some entity, which have an access to the shared state.

----

Server side of the queries also have alternative implementations:

* Use same point operations as defined in the current gPRC interface, but extend them with additional `shard_id` parameter. If this parameter is empty - collection should perform full search, if not - only inside the specified local shard.

* Implement special interfaces for inter-node communication and use a separate runtime for them. In this case API could also be more easily extended with additional methods (like e.g. collection replication). Protobuf definitions should be moved to the separate sub-crate for re-use.

**Comment by generall** [comment]
depends on: https://github.com/qdrant/qdrant/issues/407
